### PR TITLE
cmd/snap-seccomp: initial support for negative arguments with uid/gid caching

### PR DESCRIPTION
**Pull request by jdstrand**
This PR can land alone and is part of a series of PRs that break up https://github.com/snapcore/snapd/pull/6681. Policy updates utilizing this will come in a subsequent PR.

- cmd/snap-seccomp: use strtoll() instead of atoi()
- initial support for negative arguments
- snap-seccomp: implement uidCache and gidCache to reduce lookups

### CODE REVIEW

File: cmd/snap-seccomp/main.go [hunk 1]
```diff
@@ -465,13 +465,65 @@ func (sc *SeccompData) SetArgs(args [6]uint64) {
...
+	}
+
+	// It is, so try to parse as an int32
+	value, err := strconv.ParseInt(token, 10, 32)
+	if err != nil {
+		return 0, err
+	}
+
+	// convert the int32 to uint32 then to uint64 (see above)
+	return uint64(uint32(value)), nil
```
Comment (bboozzoo) [inline cmd/snap-seccomp/main.go]: Nice!

File: cmd/snap-seccomp/main.go [hunk 2]
```diff
@@ -538,15 +591,26 @@ func parseLine(line string, secFilter *seccomp.ScmpFilter) error {
...
 		if err != nil {
 			return fmt.Errorf("cannot parse token %q (line %q)", arg, line)
 		}
 
+		// For now only support EQ with negative args. If change this,
+		// be sure to adjust readNumber accordingly and use libseccomp
+		// carefully.
+		if syscallsWithNegArgsMaskHi32[syscallName] {
+			if cmpOp != seccomp.CompareEqual {
+				return fmt.Errorf("cannot parse token %q (line %q)", arg, line)
```
Comment (bboozzoo) [inline cmd/snap-seccomp/main.go]: Perhaps extend the error message to something like: `cannot parse token %q (line %q): unsupported comparison`?

File: cmd/snap-seccomp/main_test.go [hunk 3]
```diff
@@ -137,11 +137,18 @@ var seccompSyscallRunnerContent = []byte(`
...
-    for (int i = 0; i < 7; i++)
-        l[i] = atoi(argv[i + 1]);
+    uint32_t l[7];
+    int syscall_ret, ret = 0;
+    for (int i = 0; i < 7; i++) {
+        errno = 0;
+        l[i] = strtoll(argv[i + 1], NULL, 10);
+	// exit '11' let's us know strtoll failed
+        if (errno != 0)
+            syscall(SYS_exit, 11, 0, 0, 0, 0, 0);
```
Comment (zyga) [inline cmd/snap-seccomp/main_test.go]: Nice

File: cmd/snap-seccomp/main_test.go [hunk 4]
```diff
@@ -521,6 +539,10 @@ func (s *snapSeccompSuite) TestCompileBadInput(c *C) {
 		{"setpriority 1-", `cannot parse line: cannot parse token "1-" .*`},
 		{"setpriority 1\\ 2", `cannot parse line: cannot parse token "1\\\\" .*`},
 		{"setpriority 1\\n2", `cannot parse line: cannot parse token "1\\\\n2" .*`},
+		// 1 bigger than uint32
+		{"chown 0 4294967296", `cannot parse line: cannot parse token "4294967296" .*`},
+		// 1 bigger than int32
```
Comment (zyga) [inline cmd/snap-seccomp/main_test.go]:
```suggestion
		// 1 smaller than int32
```

**Comment by chipaca** [comment]
When you address those nits, please also merge master (new syscalls / seccomp fixes)

**Comment by jdstrand** [comment]
Thanks for the reviews!

**Pull request merged**